src/Build/BackEnd/Components/RequestBuilder/RequestBuilder.cs

```diff
@@ -823,14 +823,17 @@ private async Task BuildAndReport()
             catch (Exception ex)
             {
                 thrownException = ex;
-
-                if (ExceptionHandling.IsCriticalException(ex))
+                if (ex is BuildAbortedException)
+                {
+                    // The build was likely cancelled. We do not need to log an error in this case.
+                }
+                else if (_projectLoggingContext is null)
+                {
+                    _nodeLoggingContext.LogError(BuildEventFileInfo.Empty, "UnhandledMSBuildError", ex.Message, ex.StackTrace);
+                }
+                else
                 {
-                    // This includes InternalErrorException, which we definitely want a callstack for.
-                    // Fortunately the default console UnhandledExceptionHandler will log the callstack even
-                    // for unhandled exceptions thrown from threads other than the main thread, like here.
-                    // Less fortunately NUnit doesn't.
-                    throw;
```

Comment:
It never crashed very hard—even with critical exceptions, we still weren't seeing a stack or anything if it happened at the wrong time. It basically just jumped up the stack a little faster.

Reply:
My MSBuild senses are tingling with this change, but I have no evidence to push against it. It'll fail in the same way as an InvalidProjectFileException, which can't be that bad right?

Reply:
I think we should filter the catch on `IsCriticalException()`, as we do elsewhere. If we've hit an OOM or something the best thing we can do is crash immediately and let Watson tell us about it; ideally we could log but we want to see the crash in the place it came from, rather than in logging.

Reply:
How would Watson get anything? I walked through this case with KirillOsenkov and showed him that even before, a critical exception just disappeared without a trace.

Reply:
Ok, then that's bad and we should fix it. Why isn't our UnhandledException handler firing? Is it getting correctly instantiated?

Reply:
I inserted an OutOfMemoryException in LazyItemEvaluator with this change, and it logged it properly. That seems good.

Reply:
How would an UnhandledException handler know about the exception? I'm slightly hazy on this, but it seems like crash the thread hard, and essentially nothing escapes. I should clarify that there's a point when it writes to a debug file in Temp, but it never says what that file is, so there's a 0.01% chance that I'd find it randomly without having paused the debugger to see exactly which file it is.